packages/transaction-messages/src/use-address-lookup-tables.ts

```diff
@@ -0,0 +1,86 @@
+import { Address } from '@solana/addresses';
+import { AccountRole, IAccountLookupMeta, IInstruction, isSignerRole } from '@solana/instructions';
+
+import { AddressesByLookupTableAddress } from './addresses-by-lookup-table-address';
+import { ITransactionMessageWithFeePayer } from './fee-payer';
+import { TransactionMessage } from './transaction-message';
+
+type Mutable<T> = {
+    -readonly [P in keyof T]: T[P];
+};
+
+// Look up the address in lookup tables, return a lookup meta if it is found in any of them
+function findAddressInLookupTables(
+    address: Address,
+    role: AccountRole.READONLY | AccountRole.WRITABLE,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): IAccountLookupMeta | undefined {
+    for (const [lookupTableAddress, addresses] of Object.entries(addressesByLookupTableAddress)) {
+        for (let i = 0; i < addresses.length; i++) {
+            if (address === addresses[i]) {
+                return {
+                    address,
+                    addressIndex: i,
+                    lookupTableAddress: lookupTableAddress as Address,
+                    role,
+                };
+            }
+        }
+    }
+}
+
+type TransactionMessageNotLegacy = Exclude<TransactionMessage, { version: 'legacy' }>;
+
+export function useAddressLookupTablesForTransactionMessage<
+    TTransactionMessage extends TransactionMessageNotLegacy = TransactionMessageNotLegacy,
+>(
+    transactionMessage: TTransactionMessage,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): TTransactionMessage {
+    // All signers will need to be static keys, so we shouldn't replace them if they're used elsewhere in the transaction,
+    // even if they're also in a lookup table
+    const signers = new Set(
+        transactionMessage.instructions
+            .flatMap(i => (i.accounts ?? []).filter(a => isSignerRole(a.role)))
+            .map(a => a.address),
+    );
+    if ('feePayer' in transactionMessage) {
+        signers.add((transactionMessage as ITransactionMessageWithFeePayer).feePayer);
+    }
```

Comment:
So, I think you should only ever bail on a _particular_ account entry if it's a signer, instead of looking across _all_ accounts to make sure the address has never been used as a signer.

1. After optimization, someone _could_ change the fee payer, delete an instruction, or change signers. Skipping this whole entry could result in a sub-optimal account lookup set after such a modification.
2. The message compiler already [throws out LUT entries that have to sign](https://github.com/solana-labs/solana-web3.js/blob/9239e6ec972b4de9f0d15b197fbef1d2871759d9/packages/transaction-messages/src/compile/accounts.ts#L120-L124).

So if you have two instructions, each with a reference to address `ABC`, one a signer and one not, I think you should leave the signer entry alone, and replace the non-signer one with an `IAccountLookupMeta` _just_ on the off chance that the transaction is later modified or split into two or whatever in such a way that LUT entry _can_ all of a sudden be used.

Also makes this code simpler.

Reply:
Got it, that makes sense, thanks!